File: FirstComeFirstServe.cpp
// File Name: FirstComeFirstServe.cpp
// Date: December 15, 2022
// Time: 01:10:17
// Author : Sajid (git : sajidshahriar72543)

#include <bits/stdc++.h>
using namespace std;
void withoutArrival(int n);
void withArrival(int n);
int main()
{
    cout << "Enter no of processes: ";
    int n;
    cin >> n;
    cout << "Enter 1 if arrival time is given else 0: ";
    int choice;
    cin >> choice;
    if (choice == 1)
    {
        withArrival(n);
        // your code here
    }
    else
    {
        withoutArrival(n);
    }
    // return 0;
}
void withoutArrival(int n)
{
    int bt[n], wt[n], tat[n];
    cout << "Enter burst time of processes: ";
    for (int i = 0; i < n; i++)
    {
        cin >> bt[i]; // input burst time
    }
    wt[0] = 0; // waiting time of first process is 0
    for (int i = 1; i < n; i++)
    {
        wt[i] = bt[i - 1] + wt[i - 1]; // waiting time of other processes
    }
    for (int i = 0; i < n; i++)
    {
        tat[i] = bt[i] + wt[i]; // turn around time of processes
    }
    cout << "Process\tBT\tWT\tTAT" << endl;
    for (int i = 0; i < n; i++)
    {
        cout << i + 1 << "\t" << bt[i] << "\t" << wt[i] << "\t" << tat[i] << endl;
    }
    int sum = 0;
    for (int i = 0; i < n; i++)
    {
        sum += wt[i]; // sum of waiting time
    }
    cout << "Average waiting time: " << sum / n;
    sum = 0;
    for (int i = 0; i < n; i++)
    {
        sum += tat[i]; // sum of turn around time
    }
    cout << "Average turn around time: " << sum / n;
}
void withArrival(int n)
{
    int at[n], bt[n], wt[n], tat[n];
    cout << "Enter arrival time of processes: ";
    for (int i = 0; i < n; i++)
    {
        cin >> at[i]; // input arrival time
    }
    cout << "Enter burst time of processes: ";
    for (int i = 0; i < n; i++)
    {
        cin >> bt[i]; // input burst time
    }
    wt[0] = 0; // waiting time of first process is 0
    int ct[n];
    ct[0] = 0;
    for (int i = 0; i < n; i++)
    {
        wt[i] = 0;
        tat[i] = 0;
        ct[i + 1] = ct[i] + bt[i]; // completion time of processes
        wt[i] = ct[i] - at[i];     // waiting time of processes
        tat[i] = wt[i] + bt[i];    // turn around time of processes
    }
    cout << "Process\tAT\tBT\tCT\tWT\tTAT" << endl;
    for (int i = 0; i < n; i++)
    {
        cout << i + 1 << "\t" << at[i] << "\t" << bt[i] << "\t" << ct[i] << "\t" << wt[i] << "\t" << tat[i] << endl;
    }
    int sum = 0;
    for (int i = 0; i < n; i++)
    {
        sum += wt[i]; // sum of waiting time
    }
    cout << "Average waiting time: " << float(sum / n) << endl;
    sum = 0;
    for (int i = 0; i < n; i++)
    {
        sum += tat[i]; // sum of turn around time
    }
    cout << "Average turn around time: " << float(sum / n);
}

File: FirstComeFirstServe.py
n = int(input("Enter the number of processes: "))
top = 0
at = []
bt = []
ct = []
tat = []
wt = []
avgwt = 0
avgtat = 0
for i in range (n):
    inp = int(input("Enter the arrival time of process " + str(i+1) + ": "))
    at.append(inp)
    inp = int(input("Enter the burst time of process " + str(i+1) + ": "))
    bt.append(inp)
    print()

for i in range (n):
    if i==0: #first process
        top += bt[i] 
        ct.append(top) #completion time
    elif i>0: #other processes
        if at[i] > top: # if arrival time is greater than completion time
            top = at[i]+bt[i] 
            ct.append(top) #completion time
        else: #if arrival time is less than completion time
            top += bt[i]
            ct.append(top)
print("Arrival Time Calculated\n")

for i in range (n):
    temp = ct[i] - at[i]
    tat.append(temp) #turn around time
    temp = tat[i] - bt[i]
    wt.append(temp) #waiting time

print("Turn Around Time Calculated\n Waiting Time Calculated\n")

print("PID\tAT\tBT\tCT\tTAT\tWT")
for i in range (n):
    print(str(i+1) + "\t" + str(at[i]) + "\t" + str(bt[i]) + "\t" + str(ct[i]) + "\t" + str(tat[i]) + "\t" + str(wt[i]))
    avgwt += wt[i]
    avgtat += tat[i]

avgwt = avgwt/n
avgtat = avgtat/n

print("Average Waiting Time: " + str(avgwt))
print("Average Turn Around Time: " + str(avgtat))